library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java

```diff
@@ -291,6 +299,27 @@ public boolean isLoadCompleted() {
     return loadCompleted;
   }
 
+  /**
+   * Return the indexes of the first samples from each sample queue for this media chunk
+   *
+   * @return sample index {@link SampleQueue#getWriteIndex()}
+   */
+  int [] getFirstSampleIndexes() {
+    return firstSampleIndexes;
+  }
+
+  /**
+   * Set the index of the first sample written to a sample queue from this media chunk,
+   * indexed by the caller's stream index for the sample queue
+   *
+   * @param streamIndex - caller's index for the {@link SampleQueue}
+   * @param firstSampleIndex - index value to store
+   */
+  void setFirstSampleIndex(int streamIndex, int firstSampleIndex) {
+    firstSampleIndexes = Arrays.copyOf(firstSampleIndexes, streamIndex + 1);
```

Comment:
> EDIT: Note that there is indeed an issue (already sent a fix for internal review) with the thread access comments around `HlsSampleStreamWrapper.init()`. I don't think they are related to the thread issues you mentioned above, but still worth mentioning.

What's the issue number?  Because the method is definitely called from the loader thread, this happens concurrently with sample reading on the player thread so concurrent modifications can happen.

Note, of course you know even iterating an collection in one thread that is being modified in another thread will result in an ConcurrentModificationException.   I'll help address this in the issue discussion if I can.  Just let me know how best I can help.

Reply:
The exception you mentioned shouldn't be related to the init issue. Please file a fresh issue for that, it will be easier for use to debug it.